redis_dict.py

```diff
@@ -1,73 +1,190 @@
+        self._store(key, value)
 
-    def __delitem__(self, k):
-        self.redis.delete(self.namespace + k)
+    def __delitem__(self, key):
+        self.redis.delete(self._format_key(key))
 
-    def __contains__(self, k):
-        return self._get_item(k) is not None
+    def __contains__(self, key):
+        return self._load(key)[1]
+
+    def __len__(self):
+        return len(list(self._scan_keys()))
+
+    def __iter__(self):
+        self.iter = self.iterkeys()
+        return self
 
     def __repr__(self):
-        return str(self.to_dict())
+        return str(self)
 
     def __str__(self):
-        return self.__repr__()
+        return str(self.to_dict())
 
-    def __len__(self):
-        return len(self._keys())
+    def __next__(self):
+        return next(self.iter)
+
+    def next(self):
+        return self.__next__()
 
     def _scan_keys(self, search_term=''):
-        return self.redis.scan(match=self.namespace + search_term + '*')
+        return self.get_redis.scan_iter(match='{}:{}{}'.format(self.namespace, search_term, '*'))
+
+    def get(self, key, default=None):
+        found, item = self._load(key)
+        if not found:
+            return default
+        return item
+
+    def iterkeys(self):
+        """Note: for pythone2 str is needed"""
+        to_rm = len(self.namespace) + 1
+        return (str(item[to_rm:]) for item in self._scan_keys())
 
-    def _keys(self, search_term=''):
-        return self._scan_keys(search_term)[1]
+    def iter_keys(self):
+        """Deprecated: should be removed after major version change"""
+        print("Warning: deprecated method. use iterkeys instead")
+        return self.iterkeys()
+
+    def key(self, search_term=''):
+        """Note: for pythone2 str is needed"""
+        to_rm = len(self.namespace) + 1
+        cursor, data = self.get_redis.scan(match='{}:{}{}'.format(self.namespace, search_term, '*'), count=1)
+        for item in data:
+            return str(item[to_rm:])
 
     def keys(self):
-        to_rm = len(self.namespace)
-        return [item[to_rm:] for item in self._keys()]
+        return list(self.iterkeys())
+
+    def iteritems(self):
+        """Note: for pythone2 str is needed"""
+        to_rm = len(self.namespace) + 1
+        return ((str(item[to_rm:]), self._load_raw(item)) for item in self._scan_keys())
+
+    def items(self):
+        return list(self.iteritems())
+
+    def values(self):
+        return list(self.itervalues())
+
+    def itervalues(self):
+        to_rm = len(self.namespace) + 1
+        return (self[i[to_rm:]] for i in self._scan_keys())
 
     def to_dict(self):
-        to_rm = len(self.namespace)
-        return {item[to_rm:]: self._raw_get_item(item) for item in self._keys()}
+        return dict(self.items())
+
+    def clear(self):
+        with self.pipeline():
+            for key in self:
+                del(self[key])
+
+    def pop(self, key):
+        value = self[key]
+        del(self[key])
```

Comment:
When running pop you expect a return value and the and the item to be removed.
If the item is already removed because of a race condition, this method will still have the same effect.
You have a value, and the item is removed.